### PR TITLE
Bug 1302970 - update script does not patch router if name is differen…

### PR DESCRIPTION
**Pull request by brenton**
…t from default

### CODE REVIEW

**Comment by brenton** [comment]
- [ ]  Figure out a better way to pass in the pods to `with_items`
- [ ]  Update the minor upgrade playbooks
- [ ]  Add the new liveness probes to the 3.1.1 playbook

File: filter_plugins/oo_filters.py [hunk 1]
```diff
@@ -670,6 +670,34 @@ def oo_31_rpm_rename_conversion(rpms, openshift_version=None):
...
 
+    @staticmethod
+    def oo_pods_match_component(pods, deployment_type, component):
+        """ Filters a list of Pods and returns the ones matching the variant.
+        """
+        #pdb.set_trace()
+        if not isinstance(pods, list):
+            raise errors.AnsibleFilterError("failed expects to filter on a list")
+        if not isinstance(deployment_type, basestring):
+            raise errors.AnsibleFilterError("failed expects openshift_version to be a string")
```
Comment (detiber) [inline filter_plugins/oo_filters.py]: error message does not match failure.

File: filter_plugins/oo_filters.py [hunk 2]
```diff
@@ -670,6 +670,34 @@ def oo_31_rpm_rename_conversion(rpms, openshift_version=None):
...
+        if not isinstance(component, basestring):
+            raise errors.AnsibleFilterError("failed expects openshift_version to be a string")
```
Comment (detiber) [inline filter_plugins/oo_filters.py]: error message does not match failure.

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml [hunk 3]
```diff
@@ -567,46 +572,49 @@
...
 
   - name: Update deployment config to 1.0.4/3.0.1 spec
-    when: _default_router.rc == 0
+    when: _all_routers.rc == 0
     command: >
-      {{ oc_cmd }} patch dc/router -p
+      {{ oc_cmd }} patch dc/{{ item['labels']['deploymentconfig'] }} -p
       '{"spec":{"strategy":{"rollingParams":{"updatePercent":-10},"spec":{"serviceAccount":"router","serviceAccountName":"router"}}}}'
       --api-version=v1
+    with_items: (_all_routers.stdout | from_json)['items'] | oo_pods_match_component(openshift_deployment_type, 'haproxy-router') | oo_select_keys_from_list(['metadata'])
```
Comment (brenton) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: @detiber, @abutcher, I really struggled to make this do the right thing and right now there's a lot of duplicate logic.  What would be the best way to save the matching pods to another variable that could be reused?  I was probably doing something wrong at the time but the last time I tried using set_fact it was messing things up.  It that should work for a data structure like this then I can try again.

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml [hunk 4]
```diff
@@ -74,6 +74,11 @@
 
   - set_fact:
       g_new_version: "{{ g_aos_versions.curr_version.split('-', 1).0 if g_aos_versions.avail_version is none else g_aos_versions.avail_version.split('-', 1).0 }}"
+    when: openshift_pkg_version is not defined
+
+  - set_fact:
+      g_new_version: "{{ openshift_pkg_version | replace('-','') }}"
+    when: openshift_pkg_version is defined
```
Comment (brenton) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: This is just another bug fix.  Previously if you set openshift_pkg_version to something less than the latest the correct RPM would be installed but the router would be updated to the latest image.

File: filter_plugins/oo_filters.py [hunk 5]
```diff
@@ -670,6 +670,34 @@ def oo_31_rpm_rename_conversion(rpms, openshift_version=None):
...
+
+        image_prefix = 'openshift/origin-'
+        if deployment_type in ['enterprise', 'online', 'openshift-enterprise']:
+            image_prefix = 'openshift3/ose-'
+        elif deployment_type == 'atomic-enterprise':
+            image_prefix = 'aep3_beta/aep-'
```
Comment (detiber) [inline filter_plugins/oo_filters.py]: Do we want 'aep3_beta' here or just 'aep3'?

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml [hunk 6]
```diff
@@ -567,46 +572,49 @@
   - role: openshift_examples
     openshift_examples_import_command: replace
   pre_tasks:
-  - name: Check for default router
+  - name: Collect all routers
     command: >
-      {{ oc_cmd }} get -n default dc/router
-    register: _default_router
+      {{ oc_cmd }} get pods --all-namespaces -l 'router' -o json
+    register: _all_routers
```
Comment (detiber) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: Variables with leading underscores cause issues with Ansible 2.0

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml (hunk 3 above)
Comment (detiber) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: @brenton set_fact *should* work, you just might have to throw a from_yaml filter in there if it is being treated as a string.

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml (hunk 4 above)
Comment (detiber) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: gotta love those edge cases :)

File: filter_plugins/oo_filters.py (hunk 5 above)
Comment (brenton) [inline filter_plugins/oo_filters.py]: I think we'll need to take a pass to swap these out as part of a separate PR.

Comment (detiber) [inline filter_plugins/oo_filters.py]: Sounds good to me

File: playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml (hunk 3 above)
Comment (brenton) [inline playbooks/common/openshift-cluster/upgrades/v3_0_to_v3_1/upgrade.yml]: I'll try using set_fact again now that I've changed my approach a bit.  Originally I was trying to do all this without a custom filter but I couldn't get it to work right.

**Comment by brenton** [comment]
There really isn't a good way to update the liveness probes until the cli supports the other patch operations: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#patch-operations

Today only strategic merge is available from the cli which doesn't provide a way to replace a liveness probe.  Take this for example:

`oc patch dc/custom-router -p '{"spec":{"template":{"spec":{"containers":[{"name":"router","livenessProbe":{"httpGet":{"path": "/healthz", "port": 1936, "host": "localhost", "scheme": "HTTP"},"initialDelaySeconds":10,"timeoutSeconds":1}}]}}}}'
The DeploymentConfig "custom-router" is invalid.
spec.template.spec.containers[0].livenessProbe: invalid value '<*>(0xc20f13d1d0){Exec:<nil> HTTPGet:<*>(0xc20ee3d4a0){Path:/healthz Port:1936 Host:localhost Scheme:HTTP} TCPSocket:<*>(0xc20c74d400){Port:80}}', Details: exactly 1 handler type is required`

What it's doing is trying to update the liveness probe with the new httpGet probe and still leave the TCPSocket probe in place.  As a workaround I thought I could effectively erase the probe by doing this:

`oc patch dc/custom-router -p '{"spec":{"template":{"spec":{"containers":[{"name":"router","livenessProbe":{}}]}}}}'
"custom-router" patched`

However, that results in no change.  I see upstream kube talking about adding cli support for the additional patch operations so my vote is to wait until then. 

cc @detiber, @sdodson

**Comment by detiber** [comment]
@brenton would it be better to avoid the patch and just update the DC directly? or does that suffer from the same issue?

**Comment by sdodson** [comment]
I'm fine with waiting for this particular change, we'll want to evaluate future changes as some we may not be able to wait for. Depending on how fast kube adds additional patch operations perhaps we should consider including jq (https://stedolan.github.io/jq/) in our repos so we can perform advanced json modification?

**Comment by brenton** [comment]
@sdodson, I like the idea of using jq if it comes down to that.  To @detiber's point I just didn't want to deal with the additional complexity of having to carefully modify the entire json object just yet.  These liveness probes aren't actually causing a real problem right now (at least not once 3.1.1.6 is released) and I'd rather wait until we have either better access to the patch API or something like jq available.

File: playbooks/common/openshift-cluster/upgrades/v3_1_minor/pre.yml [hunk 7]
```diff
@@ -41,7 +41,7 @@
       g_new_service_name: "{{ 'origin' if deployment_type =='origin' else 'atomic-openshift' }}"
 
   - name: Determine available versions
-    script: ../files/versions.sh {{ g_new_service_name }} openshift
+    script: ../files/versions.sh {{ g_new_service_name }}
```
Comment (brenton) [inline playbooks/common/openshift-cluster/upgrades/v3_1_minor/pre.yml]: @sdodson, @detiber, this is another one of those things that I have no idea how anyone's 3.1.1 upgrade actually worked.  This needs to be fixed asap.  In the minor upgrade we only need to know the latest version of the atomic-openshift package.

**Comment by brenton** [comment]
@detiber, PTAL.  I think this is ready for QE.  I'll write up all the various test cases they should try in the BZ.

File: playbooks/common/openshift-cluster/upgrades/v3_1_minor/pre.yml (hunk 7 above)
Comment (sdodson) [inline playbooks/common/openshift-cluster/upgrades/v3_1_minor/pre.yml]: I think this attempted to handle cases where they're upgrading within 3.0.x. The `sort -r` in versions.sh should make this a non issue assuming that atomic-openshift version is higher than that of the available openshift version. I think this is safe to remove at this point so +1

Comment (brenton) [inline playbooks/common/openshift-cluster/upgrades/v3_1_minor/pre.yml]: Yeah, it's needed for 3.0.z upgrades.  Just not for 3.1.z.

**Comment by detiber** [comment]
I'm not sure we need something like jq, especially since we are talking about wrapping the client/admin commands in a module. At that point we can just manipulate the json within the module if needed.

@brenton I say :shipit: to QE

**Comment by brenton** [comment]
@detiber, That makes sense.

**Pull request merged**